### PR TITLE
Feature/nijo 1019 uploading images

### PR DESCRIPTION
**Pull request by Raibo**

### CODE REVIEW

File: Api/Services/Files/ImageFileService.cs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            _s3FolderName = options.Value.S3FolderName;
+            _agentContextService = agentContextService;
+            _edoContext = edoContext;
+            _dateTimeProvider = dateTimeProvider;
+        }
+
+
+        public async Task<Result> Add(IFormFile file)
+        {
+            var agentContext = await _agentContextService.GetAgent();
```
Comment (mingazhev) [inline Api/Services/Files/ImageFileService.cs:42]: You should pass agentContext from controller

File: Api/Services/Files/ImageFileService.cs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+        private string GetKey(int agencyId, string fileName) => $"{_s3FolderName}/{agencyId}/{fileName.ToLower()}";
+
+
+        private const int MinimumWidth = 200;
+        private const int MaximumWidth = 800;
+        private const int MinimumHeight = 200;
+        private const int MaximumHeight = 800;
+
+        private readonly List<string> _imageExtensions = new List<string>{"jpeg", "jpg"};
```
Comment (mingazhev) [inline Api/Services/Files/ImageFileService.cs:179]: Please use hashset for such things

Comment (mingazhev) [reply Api/Services/Files/ImageFileService.cs:179]: And this could be static readonly

File: Api/Services/Files/ImageFileService.cs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            return await Validate()
+                .Bind(Upload)
+                .Map(StoreLink);
+
+
+            async Task<Result> Validate()
+            {
+                return await Result.Success()
+                    .Ensure(() => file != default, "Could not get the file")
+                    .Ensure(() => _imageExtensions.Contains(Path.GetExtension(file?.FileName)?.ToLower()), "The file must have extension of a jpeg image")
```
Comment (mingazhev) [inline Api/Services/Files/ImageFileService.cs:53]: ToLowerInvariant

File: Api/Controllers/AgentControllers/AgencyImagesController.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+
+        /// <summary>
+        ///     Uploads an image
+        /// </summary>
+        [HttpPut("images")]
+        [ProducesResponseType((int) HttpStatusCode.OK)]
+        [ProducesResponseType(typeof(ProblemDetails), (int) HttpStatusCode.BadRequest)]
+        [InAgencyPermissions(InAgencyPermissions.AgencyImagesManagement)]
+        public async Task<IActionResult> Add([FromForm] IFormFile file) => OkOrBadRequest(await _imageFileService.Add(file));
```
Comment (mingazhev) [inline Api/Controllers/AgentControllers/AgencyImagesController.cs:32]: Don't think that we need such general-purpose endpoint.
For now we need a banner and logo only.
Why not loading and getting them throw separate endpoints?

@kirillta what do you think?

Comment (AlexeyShok) [reply Api/Controllers/AgentControllers/AgencyImagesController.cs:32]: Why is the HttpPut request used? If the method "Add" adds a picture then it's should be a HttpPost. Either the method is named incorrectly.

File: Api/Services/Files/ImageFileService.cs [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+                            $"Image width must be in range from {MinimumWidth} to {MaximumWidth}")
+                        .Ensure(() => imageInfo.ImageHeight >= MinimumHeight && imageInfo.ImageHeight <= MaximumHeight,
+                            $"Image height must be in range from {MinimumWidth} to {MaximumWidth}");
+                }
+            }
+
+
+            async Task<Result<string>> Upload()
+            {
+                await using var stream = file.OpenReadStream();
```
Comment (AlexeyShok) [inline Api/Services/Files/ImageFileService.cs:83]: Problems may arise due to incorrect work with the file stream. The stream is opened for reading twice: during validation and then when loaded into S3. Possible slowdown or crash when rereading from the file stream. This solution needs to be tested thoroughly.

File: Api/Services/Files/ImageFileService.cs (hunk 1 above)
Comment (kirillta) [reply Api/Services/Files/ImageFileService.cs:42]: Why, isn't that an unnecessary dependency?

File: Api/Controllers/AgentControllers/AgencyImagesController.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+
+        /// <summary>
+        ///     Gets all uploaded images for the current agency
+        /// </summary>
+        [HttpGet("images")]
+        [ProducesResponseType(typeof(List<SlimUploadedImage>), (int) HttpStatusCode.OK)]
+        [ProducesResponseType(typeof(ProblemDetails), (int) HttpStatusCode.BadRequest)]
```
Comment (kirillta) [inline Api/Controllers/AgentControllers/AgencyImagesController.cs:40]: This method returns only 200

File: Api/Models/Agencies/SlimUploadedImage.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+        /// <summary>
+        /// File name of the image
+        /// </summary>
+        public string FileName { get; }
+
+        /// <summary>
+        /// Download link of the image
+        /// </summary>
+        public string Link { get; }
```
Comment (kirillta) [inline Api/Models/Agencies/SlimUploadedImage.cs:25]: If I remember right we name it `Url` in other places @AlexeyShok

**Review by kirillta — CHANGES_REQUESTED** [review]

File: Api/Models/Agencies/SlimUploadedImage.cs (hunk 7 above)
Comment (AlexeyShok) [reply Api/Models/Agencies/SlimUploadedImage.cs:25]: Yes, we named it Url in hiroshima.

File: Api/Controllers/AgentControllers/AgencyImagesController.cs (hunk 4 above)
Comment (Raibo) [reply Api/Controllers/AgentControllers/AgencyImagesController.cs:32]: @AlexeyShok , HttpPut is used due to the fact that new picture with the same name replaces the old one.
If you think that method named incorrectly, suggest a better name.

File: Api/Services/Files/ImageFileService.cs (hunk 1 above)
Comment (Raibo) [reply Api/Services/Files/ImageFileService.cs:42]: In previous review I was told to move agentContext from controller to service. 
Here I am told to move it from service to controller.
This is strange.

File: Api/Services/Files/ImageFileService.cs (hunk 5 above)
Comment (Raibo) [reply Api/Services/Files/ImageFileService.cs]: Changed to save bytes and memory stream

File: Api/Models/Agencies/SlimUploadedImage.cs (hunk 7 above)
Comment (Raibo) [reply Api/Models/Agencies/SlimUploadedImage.cs]: Renamed

File: Api/Controllers/AgentControllers/AgencyImagesController.cs (hunk 4 above)
Comment (Raibo) [reply Api/Controllers/AgentControllers/AgencyImagesController.cs:32]: @mingazhev Why not. So, `.../banner` and `.../logo` PUT/GET endpoints?

File: Api/Services/Files/ImageFileService.cs (hunk 1 above)
Comment (mingazhev) [reply Api/Services/Files/ImageFileService.cs:41]: I told so? Maybe I was mistaken

File: Api/Controllers/AgentControllers/AgencyImagesController.cs (hunk 4 above)
Comment (mingazhev) [reply Api/Controllers/AgentControllers/AgencyImagesController.cs:32]: Yes, separate endpoints will be better. Del will be good too

File: Api/Services/Files/ImageFileService.cs (hunk 1 above)
Comment (mingazhev) [reply Api/Services/Files/ImageFileService.cs:41]: Could you give a link?

Comment (mingazhev) [reply Api/Services/Files/ImageFileService.cs:41]: And Kirill is right, could not find where this dependency used now

**Review by mingazhev — COMMENTED** [review]
Need to change the endpoints logic

File: Api/Services/Files/ImageFileService.cs (hunk 1 above)
Comment (Raibo) [reply Api/Services/Files/ImageFileService.cs:41]: Here is the link
https://github.com/happy-travel/edo/pull/283#discussion_r520647510

File: Api/Controllers/AgentControllers/AgencyImagesController.cs (hunk 4 above)
Comment (AlexeyShok) [reply Api/Controllers/AgentControllers/AgencyImagesController.cs:32]: If HttpPut remains, then I would call method AddOrReplace or Replace, but not Add

File: Api/Controllers/AgentControllers/AgencyImagesController.cs [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+
+        /// <summary>
+        ///     Uploads an image for banner
+        /// </summary>
+        [HttpPut("banner")]
+        [ProducesResponseType((int)HttpStatusCode.OK)]
+        [ProducesResponseType(typeof(ProblemDetails), (int)HttpStatusCode.BadRequest)]
+        [InAgencyPermissions(InAgencyPermissions.AgencyImagesManagement)]
+        public async Task<IActionResult> AddOrReplaceBanner([FromForm] IFormFile file)
```
Comment (mingazhev) [inline Api/Controllers/AgentControllers/AgencyImagesController.cs:35]: SetBanner maybe?

**Review by mingazhev — APPROVED** [review]
With a small comment

**Review by kirillta — APPROVED** [review]

File: Api/Controllers/AgentControllers/AgencyImagesController.cs (hunk 4 above)
Comment (Raibo) [reply Api/Controllers/AgentControllers/AgencyImagesController.cs]: I also considered AddOrReplace, but after discussing with Vadim, renamed to Set

**Pull request merged**